runtimes/pythonrt/worker_grpc_handler.go

```diff
@@ -0,0 +1,283 @@
+package pythonrt
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"sync"
+
+	pb "go.autokitteh.dev/autokitteh/proto/gen/go/autokitteh/remote/v1"
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type workerGRPCHandler struct {
+	pb.UnimplementedWorkerServer
+
+	runnerIDsToRuntime map[string]*pySvc
+	mu                 *sync.Mutex
+}
+
+var (
+	w = workerGRPCHandler{
+		runnerIDsToRuntime: map[string]*pySvc{},
+		mu:                 new(sync.Mutex),
+	}
+	Server *grpc.Server
+)
+
+// GRPC Server Handling
+func newInterceptor(log *zap.Logger) func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+	fn := func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+		log.Info("call", zap.String("method", info.FullMethod))
+
+		return handler(ctx, req)
+	}
+
+	return fn
+}
+
+func ConfigureWorkerGRPCHandler(l *zap.Logger) {
+	srv := grpc.NewServer(grpc.UnaryInterceptor(newInterceptor(l)))
+	pb.RegisterWorkerServer(srv, &w)
+	Server = srv
+}
+
+func addRunnerToServer(runnerID string, svc *pySvc) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	_, ok := w.runnerIDsToRuntime[runnerID]
+	if ok {
+		return errors.New("already registered")
+	}
+	w.runnerIDsToRuntime[runnerID] = svc
+	return nil
+}
+
+func removeRunnerFromServer(runnerID string) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	_, ok := w.runnerIDsToRuntime[runnerID]
+	if !ok {
+		return errors.New("unknown runner id")
+	}
+	delete(w.runnerIDsToRuntime, runnerID)
+	return nil
+}
+
+// GRPC Handlers
+// TODO: call temporal to verify workflow is still active ?
+// TODO: add runner ID to health check so we can verify it
+func (s *workerGRPCHandler) Health(ctx context.Context, req *pb.HealthRequest) (*pb.HealthResponse, error) {
+	return &pb.HealthResponse{}, nil
+}
+
+func (s *workerGRPCHandler) Log(ctx context.Context, req *pb.LogRequest) (*pb.LogResponse, error) {
+	if req.Level == "" {
+		return nil, status.Error(codes.InvalidArgument, "empty level")
+	}
+
+	w.mu.Lock()
+	runner, ok := w.runnerIDsToRuntime[req.RunnerId]
+	w.mu.Unlock()
+	if !ok {
+		return &pb.LogResponse{
+			Error: "Unknown runner id",
+		}, nil
+	}
+
+	runner.channels.log <- req
+	return &pb.LogResponse{}, nil
+}
+
+func (s *workerGRPCHandler) Print(ctx context.Context, req *pb.PrintRequest) (*pb.PrintResponse, error) {
+	w.mu.Lock()
+	runner, ok := w.runnerIDsToRuntime[req.RunnerId]
+	w.mu.Unlock()
+	if !ok {
+		return &pb.PrintResponse{
+			Error: "Unknown runner id",
+		}, nil
+	}
+
+	runner.channels.print <- req
+	return &pb.PrintResponse{}, nil
+}
+
+func (s *workerGRPCHandler) Done(ctx context.Context, req *pb.DoneRequest) (*pb.DoneResponse, error) {
+	w.mu.Lock()
+	runner, ok := w.runnerIDsToRuntime[req.RunnerId]
+	w.mu.Unlock()
+	if !ok {
+		return &pb.DoneResponse{}, nil
```

Comment:
No error. Also set return gRPC errors. For example:

```
return nil, status.Errorf(codes.Internal, "new function value: %s", err)
```

See `svc.go` in my branch.